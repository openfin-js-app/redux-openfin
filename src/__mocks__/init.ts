import {
    ChannelType, IConfig, IInitState,
} from "../init";

export const initState:IInitState ={
    fin:{},
    store:{
        dispatch:jest.fn(),
        getState:()=>({}),
        subscribe:jest.fn(),
        replaceReducer:jest.fn(),
    },
};