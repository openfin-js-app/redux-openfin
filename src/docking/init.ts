import {FinWindow} from '../GlobalTypes';

import toLegacyFinWindow from '../utils/toLegacyFinWindow';

import {
    IDockingOptions, IRectangle
} from './DockingType'

import {
    getAppId, requestMonitorInfo,
    parsePositiveInt, parseOpacity,
} from './DockingUtil';

import LocalStoragePersistence from './LocalStoragePersistence';
import DockingManager from './DockingManager';


export type State = {
    monitors:IRectangle[],
    persistenceService:LocalStoragePersistence,
    options:IDockingOptions,
    dockingManager:DockingManager,
}

const DOCKING_NAMESPACE_PREFIX = 'AL90_REDUX_OPENFIN_DOCKING::';

export const initState:State = {
    monitors:[],
    persistenceService: null,
    options:null,
    dockingManager:null,
};

const defaultDockingOptions:IDockingOptions = {
    range: 20,
    spacing: 0,
    undockOffsetX: 25,
    undockOffsetY: 25,
    movingOpacity: 0.6,
    snappedMovingOpacity: 0.8,
    snappedTargetOpacity: 1,
    dockableToOthers: true,
    unregisterOnClose: true,
};

async function initMonitorInfo(fin) {
    const monitors:IRectangle[] = await requestMonitorInfo(fin);
    for (const monitorInfo of monitors) {
        // add to monitors array, rather than replacing the ref, in case window already initialised with ref
        initState.monitors.push(monitorInfo);
    }
    
}

export default (fin,mainFinWindow:FinWindow,dockingOptions:Partial<IDockingOptions>)=>{

    const options = Object.assign(defaultDockingOptions,dockingOptions);

    initMonitorInfo(fin);

    options.range = parsePositiveInt(options.range,defaultDockingOptions.range);
    options.spacing = parsePositiveInt(options.spacing,defaultDockingOptions.spacing);
    options.undockOffsetX = parsePositiveInt(options.undockOffsetX,defaultDockingOptions.undockOffsetX);
    options.undockOffsetY = parsePositiveInt(options.undockOffsetY,defaultDockingOptions.undockOffsetY);
    options.movingOpacity = parseOpacity(options.movingOpacity,defaultDockingOptions.movingOpacity);
    options.snappedMovingOpacity = parseOpacity(options.snappedMovingOpacity,defaultDockingOptions.snappedMovingOpacity);
    options.snappedTargetOpacity = parseOpacity(options.snappedTargetOpacity,defaultDockingOptions.snappedTargetOpacity);

    initState.persistenceService = new LocalStoragePersistence(DOCKING_NAMESPACE_PREFIX+getAppId(fin));

    initState.options = options;

    initState.dockingManager = new DockingManager(options);

    // register global unDockAll hot key
    fin.desktop.GlobalHotkey.register("Shift+CommandOrControl+U",()=>{
        if (initState.dockingManager){
            initState.dockingManager.undockAll();
        }
    });

    initState.dockingManager.register(toLegacyFinWindow(mainFinWindow),false)

}