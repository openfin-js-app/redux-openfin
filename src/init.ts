import { Store } from 'redux';

import registerDefaultListener from './event/registerDefaultListener';

export interface IConfig{
    ignoreStore?:boolean
}

export interface IInitState {
    fin:any,
    store:Store<any>,
}

export const initState:IInitState ={
    fin:null,
    store:null,
};

export default (fin: any, config:IConfig, store?: Store<any>)=>{

    if (store){
        registerDefaultListener(fin,store);
    }
    initState.fin = fin;

    if (config.ignoreStore){
        initState.store = null;
    }else{
        initState.store = store;
    }

}