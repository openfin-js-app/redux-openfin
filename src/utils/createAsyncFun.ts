import {Action} from 'redux-actions';

import { initState } from '../init';
import {FIN_NOT_INJECTED_MSG, LIB_REDUX_DISPATCH_FIELD_NAME } from "../GlobalTypes";

declare const window:any;

/**
 * async wrapper creator for api v1
 *
 * @deprecated
 * switch to use wrapAsyncFun instead
 *
 * @param action
 * @param ERROR_MSG
 * @param resActionCreator
 * @param finCb
 */
export default function createAsyncFun<RequestPayloadType,ResponsePayloadType,>(
        action:Action<RequestPayloadType>,
        ERROR_MSG:string,
        resActionCreator:(options:ResponsePayloadType)=>Action<ResponsePayloadType>,
        finCb:(
            fin:any,
            action:Action<RequestPayloadType>,
            resActionCreator:(options:ResponsePayloadType)=>Action<ResponsePayloadType>,
            succCb:(action:Action<ResponsePayloadType>)=>void,
            errCb:(e:Error|string)=>void
        )=>void
):Promise<Action<ResponsePayloadType>> {

    const libReduxDispatch =
        (action[LIB_REDUX_DISPATCH_FIELD_NAME] && window[action[LIB_REDUX_DISPATCH_FIELD_NAME]])?
            window[action[LIB_REDUX_DISPATCH_FIELD_NAME]]:
            void 0;

    return new Promise<Action<ResponsePayloadType>>((resolve,reject)=>{

        const succCb = (action:Action<ResponsePayloadType>)=>{
            if (libReduxDispatch){
                libReduxDispatch(action)
            }else if (initState.store){
                initState.store.dispatch(action);
            }
            resolve(action);;
        };
        const errCb = (e:Error|string)=>{
            const errMsg:string = e && (e as Error).message
                ? (e as Error).message
                : e?e as string:ERROR_MSG;
            const error = new Error(errMsg);
            const rejectAction = resActionCreator({
                name:'Error',
                error
            } as any);
            if (libReduxDispatch){
                libReduxDispatch(action)
            }else if (initState.store){
                initState.store.dispatch(rejectAction);
            }
            reject(rejectAction.error)
        };

        if (initState.fin){
            finCb(initState.fin,action,resActionCreator,succCb,errCb);
        }else{
            if (initState.store){
                const rejectAction = resActionCreator({
                    name:'Error',
                    error:new Error(FIN_NOT_INJECTED_MSG),
                } as any );
                initState.store.dispatch(rejectAction);
                reject(rejectAction.error)
            }else{
                reject(new Error(FIN_NOT_INJECTED_MSG))
            }
        }

    });

}