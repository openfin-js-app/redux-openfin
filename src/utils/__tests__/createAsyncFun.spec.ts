jest.mock('../../init');
import {initState} from '../../init';

import createAsyncFun from '../createAsyncFun';

describe('createAsyncFun util',()=>{

    afterEach(()=>{
        jest.restoreAllMocks();
    });

    it('async resolved',async ()=>{

        const response = await createAsyncFun<any,any>(
            {
                type:'action_name',
                payload:{}
            },
            'err_msg',
            (options:any):any=>({type:'action_name_res',payload:options}),
            (fin,action,resActionCreator,succCb,errCb)=>{
                succCb({
                    type:'res_action_name',
                    payload:{}
                })
            }
        );

        expect(response).toMatchSnapshot();
        expect(initState.store.dispatch).toMatchSnapshot();

    })

    it('async rejected with reason str',async ()=>{

        try{
            await createAsyncFun<any,any>(
                {
                    type:'action_name',
                    payload:{}
                },
                'err_msg',
                (options:any):any=>({type:'action_name_res',payload:options}),
                (fin,action,resActionCreator,succCb,errCb)=>{
                    errCb('err')
                }
            );
        }catch (e) {
            expect(e).toMatchSnapshot();
        }
        expect(initState.store.dispatch).toMatchSnapshot();

    })

    describe('fin not injected',()=>{

        beforeEach(()=>{
            initState.fin = null;
        })

        afterEach(()=>{
            initState.fin = {};
        })

        it('throw exception by default',async ()=>{
            try{
                await createAsyncFun<any,any>(
                    {
                        type:'action_name',
                        payload:{}
                    },
                    'err_msg',
                    (options:any):any=>({type:'action_name_res',payload:options}),
                    (fin,action,resActionCreator,succCb,errCb)=>{
                        succCb({
                            type:'res_action_name',
                            payload:{}
                        })
                    }
                );
            }catch (e) {
                expect(e).toMatchSnapshot();
            }
            expect(initState.store.dispatch).toMatchSnapshot();
        })
    })

    describe('fin & store not injected',()=>{

        const oldStore = initState.store;

        beforeEach(()=>{
            initState.fin = null;
            initState.store = null;
        })

        afterEach(()=>{
            initState.fin = {};
            initState.store = oldStore;
        })

        it('throw exception by default',async ()=>{
            try{
                await createAsyncFun<any,any>(
                    {
                        type:'action_name',
                        payload:{}
                    },
                    'err_msg',
                    (options:any):any=>({type:'action_name_res',payload:options}),
                    (fin,action,resActionCreator,succCb,errCb)=>{
                        succCb({
                            type:'res_action_name',
                            payload:{}
                        })
                    }
                );
            }catch (e) {
                expect(e).toMatchSnapshot();
            }
        })
    })

});