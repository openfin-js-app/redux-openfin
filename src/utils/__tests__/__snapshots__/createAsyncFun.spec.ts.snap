// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAsyncFun util async rejected with reason str 1`] = `undefined`;

exports[`createAsyncFun util async rejected with reason str 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "type": "res_action_name",
      },
    ],
    Array [
      Object {
        "payload": Object {
          "error": [Error: err],
          "name": "Error",
        },
        "type": "action_name_res",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`createAsyncFun util async resolved 1`] = `
Object {
  "type": "res_action_name",
}
`;

exports[`createAsyncFun util async resolved 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "type": "res_action_name",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`createAsyncFun util fin & store not injected throw exception by default 1`] = `[Error: fin handler is not injected or initialized]`;

exports[`createAsyncFun util fin not injected throw exception by default 1`] = `undefined`;

exports[`createAsyncFun util fin not injected throw exception by default 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "type": "res_action_name",
      },
    ],
    Array [
      Object {
        "payload": Object {
          "error": [Error: err],
          "name": "Error",
        },
        "type": "action_name_res",
      },
    ],
    Array [
      Object {
        "payload": Object {
          "error": [Error: fin handler is not injected or initialized],
          "name": "Error",
        },
        "type": "action_name_res",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
